useragent/patterns.go

```diff
@@ -14,25 +14,28 @@ const (
 
 var regexpSemVer = regexp.MustCompile(`^` + semVerCore + semVerPrerelease + semVerBuildmetadata + `$`)
 
-var regexpAlphanum = regexp.MustCompile(`^[0-9A-Za-z_-]+$`)
-
 func matchSemVer(s string) error {
 	if regexpSemVer.MatchString(s) {
 		return nil
 	}
 	return fmt.Errorf("invalid semver string: %s", s)
 }
 
-func matchAlphanum(s string) error {
-	if regexpAlphanum.MatchString(s) {
-		return nil
-	}
-	return fmt.Errorf("invalid alphanumeric string: %s", s)
+// Alphanumeric characters, hyphen, underscore, and period. This is the subset of
+// characters that we allow in user agent keys and values. This is to ensure that
+// downstream applications can correctly parse the full user agent header.
+//
+// NOTE: HTTP headers in general only work well with ASCII characters. see:
+// https://stackoverflow.com/questions/4400678/what-character-encoding-should-i-use-for-a-http-header
+var validChars = `0-9A-Za-z_\-\.`
```

Comment:
Please initialize the regexps you wish to use as globals as well. This string (which can be a const) is not useful outside of regexps. It can also include the brackets because it is always a character class.

The regexp compile adds overhead for every call (multiple calls per request) that we can amortize by initializing it only once.

Reply:
removed